### PR TITLE
Get thor mem fix 0.14.1

### PR DESCRIPTION
**Pull request by mansfieldbitter**

### CODE REVIEW

File: hpycc/save.py [hunk 1]
```diff
@@ -46,13 +62,15 @@ def save_output(connection, script, path_or_buf=None, syntax_check=True,
     return result.to_csv(path_or_buf=path_or_buf, **kwargs)
 
 
-def save_outputs(connection, script, directory=".", filenames=None,
-                 prefix=None, syntax_check=True, delete_workunit=True,
+def save_outputs(connection, script, directory=".", overwrite=True,
```
Comment (cqkh42) [inline hpycc/save.py]: I'm happy to have this debate about whether we should change this, but I don't think this is the forum/ issue for it. Do you want to open up a new issue/ pull request for these changes?

File: hpycc/get.py [hunk 2]
```diff
@@ -9,20 +9,20 @@
 ---------
 - `get_output` -- Return the first output of an ECL script.
 - `get_outputs` -- Return all outputs of an ECL script.
-- `get_logical_file` -- Return the contents of a logical file.
+- `get_logical_file` -- Deprecated
```
Comment (cqkh42) [inline hpycc/get.py]: We can delete this in this version.

File: hpycc/get.py [hunk 3]
```diff
@@ -256,8 +257,8 @@ def get_logical_file(*args, **kwargs):
                       "instead.")
 
 
-def get_thor_file(connection, thor_file, max_workers=15, chunk_size=10000,
-                  max_attempts=3, max_sleep=10, dtype=None):
+def get_thor_file(connection, thor_file, max_workers=10, chunk_size='auto', max_attempts=3, max_sleep=60,
+                  min_sleep=50, dtype=None):
```
Comment (cqkh42) [inline hpycc/get.py]: what happens if we don't deal with min sleep here. I don't really like it and I'm not sure I want to introduce it. Can we create  a new issue/ pull request for it?

File: hpycc/get.py [hunk 4]
```diff
@@ -321,66 +326,56 @@ def get_thor_file(connection, thor_file, max_workers=15, chunk_size=10000,
...
-
-    # if there are no rows to go and get, we should return an empty dataframe
-    if not num_rows:
-        cols = [c.name for c in schema]
-        return pd.DataFrame(columns=cols)
+    resp = connection.get_chunk_from_hpcc(thor_file, 0, 1, max_attempts, max_sleep, min_sleep)
+    try:
+        wuresultresponse = resp["WUResultResponse"]
+        schema_str = wuresultresponse["Result"]["XmlSchema"]["xml"]
+        schema = parse_schema_from_xml(schema_str, dtype)
```
Comment (cqkh42) [inline hpycc/get.py]: I think this dtype bit shouldn't be dealt with here, and should just be overridden if manual. It keeps the parse schema function cleaner

File: hpycc/connection.py [hunk 5]
```diff
@@ -21,20 +21,45 @@
...
+
+
+def _make_thorname_html(logical_file):
+    """
+    quote_plus in urllib is so badly named I made this function.
+    """
+    return quote_plus(logical_file)
+
 
 def check_ecl_cmd(cmd='ecl'):
```
Comment (cqkh42) [inline hpycc/connection.py]: I really like this function, but again. Is it the scope of this pull request? I feel it should be a new issue

File: hpycc/connection.py [hunk 6]
```diff
@@ -21,20 +21,45 @@
...
+from simplejson.errors import JSONDecodeError as simpleJSONDecodeError
 
 from hpycc.utils.parsers import parse_wuid_from_failed_response, \
     parse_wuid_from_xml
 from hpycc import delete
 
+lock = threading.Lock()
+
+
+def _make_thorname_html(logical_file):
```
Comment (cqkh42) [inline hpycc/connection.py]: do we need this as a function?

File: hpycc/connection.py [hunk 7]
```diff
@@ -317,15 +351,64 @@ def run_url_request(self, url, max_attempts, max_sleep):
                 return r
             except (HTTPError, ValueError) as e:
                 attempts += 1
-                time_to_sleep = random.randint(0, max_sleep)
+                time_to_sleep = random.randint(min_sleep, max_sleep)
                 sleep(time_to_sleep)
                 if attempts == max_attempts:
                     raise RetryError(e)
 
+    def get_chunk_from_hpcc(self, logical_file, start_row, n_rows, max_attempts,
```
Comment (cqkh42) [inline hpycc/connection.py]: how do this and get-logical_file_chunk work together? do we need both?

File: hpycc/utils/parsers.py [hunk 8]
```diff
@@ -141,10 +140,7 @@ def parse_wuid_from_xml(result):
     return wuid.group(0)
 
 
-Schema = namedtuple("Schema", ["name", "is_set", "type"])
-
-
-def parse_schema_from_xml(xml):
+def parse_schema_from_xml(xml, dtype):
```
Comment (cqkh42) [inline hpycc/utils/parsers.py]: I don't like taking in dtype, it shouldn't be the role of this function

File: hpycc/utils/parsers.py [hunk 9]
```diff
@@ -153,22 +149,43 @@ def parse_schema_from_xml(xml):
...
+
+    # Check that no weird columns have been passed
+    if isinstance(dtype, dict) and any([dtype_col not in schema_out.keys() for dtype_col in dtype.keys()]):
+        raise KeyError('Not all dtype columns exist in the logical file!\nFound: %s\nGiven: %s' %
+                       (schema_out.keys(), dtype))
+
+    return schema_out
+
+
+def _determine_dtype(dtype, name, child):
```
Comment (cqkh42) [inline hpycc/utils/parsers.py]: I dont get why we need this

File: hpycc/get.py (hunk 3 above)
Comment (cqkh42) [reply hpycc/get.py]: When I say here, I mean in this pull request

File: hpycc/save.py (hunk 1 above)
Comment (mansfieldbitter) [reply hpycc/save.py]: I agree we should discuss but I'm going to be an arse and push this out as it makes other elements of our code easier to use (the change occurred once I'd spent an hour cursing over this link not existing while testing a solution!) and, to my mind, is better as it maintains the link between ECL and our output names. We can revert later -if- when I lose.

File: hpycc/get.py (hunk 3 above)
Comment (mansfieldbitter) [reply hpycc/get.py]: Removed. min_sleep is no longer passed and is instead
`math.ceil(max_sleep*0.75)`

File: hpycc/get.py (hunk 2 above)
Comment (mansfieldbitter) [reply hpycc/get.py]: Removed with extreme prejudice.

File: hpycc/connection.py (hunk 5 above)
Comment (mansfieldbitter) [reply hpycc/connection.py]: Perhaps but function names were highlighted as a limitation  in the previous pull request, hence this addition. This was a bug encountered while developing this solution (it caused tests to fail) so was added here.

File: hpycc/connection.py (hunk 7 above)
Comment (mansfieldbitter) [reply hpycc/connection.py]: This part of get-logical_file_chunk is called from more than one place, hence breaking it out. It's used to obtain chunks for data, and also for obtaining the schema. Happy to put it somewhere else but felt nicer than splitting out code. Thoughts?

File: hpycc/save.py (hunk 1 above)
Comment (mansfieldbitter) [reply hpycc/save.py]: After careful discussion, we've decided to halt this problem by just disabling the blasted thing.

File: hpycc/utils/parsers.py (hunk 9 above)
Comment (mansfieldbitter) [reply hpycc/utils/parsers.py]: Popped out into a seperate function.

Comment (mansfieldbitter) [reply hpycc/utils/parsers.py]: and removed the stuff from schema parsing.

File: hpycc/utils/parsers.py (hunk 8 above)
Comment (mansfieldbitter) [reply hpycc/utils/parsers.py]: removed the dtype stuff from here and added to a new function which is called later.

**Comment by mansfieldbitter** [comment]
Changes resolved but several commits involved. Closing this request and re-merging based on current state.

**Pull request closed without merging**